### ISSUE TITLE
Boolean attributes

### ISSUE THREAD

**Issue by breckenedge** [Theorem/coercive#8]
Have y'all considered adding boolean attributes to this library? It's lovely that a JSON use case will require no actual conversion, but right now there's no way to "passthrough" a boolean attribute, right?

**Comment by tonchis** [comment]
Good point, we don't have a `boolean` coercion function. I'll take a stab at it!

**Comment by tonchis** [comment]
Before I go implementing, what do you think of something like this?

```ruby
module Foo
  attribute :foo, boolean, required
end

# Only `true` or `false` allowed, anything else will throw `not_valid`
Foo.call("foo" => true)
# => {"foo" => true}
```

Like `boolean`, but `"true"` or `"false"` are also allowed, anything else will throw `not_valid`.
```ruby
module Foo
  attribute :foo, boolean(format: :string), required
end

Foo.call("foo" => true)
# => {"foo" => true}

# Should this be case sensitive?
Foo.call("foo" => "true")
# => {"foo" => true}
```

Like `boolean`, but `1`, `"1"`, `0` or `"0"` are also allowed, anything else will throw `not_valid`.
```ruby
module Foo
  attribute :foo, boolean(format: :integer), required
end

Foo.call("foo" => 1)
# => {"foo" => true}

Foo.call("foo" => "1")
# => {"foo" => true}
```

**Comment by breckenedge** [comment]
I like the idea of casting with a default collection. Here are the collections that I've used in the past:

```ruby
[true, 1, '1', 't', 'true', 'on']
[false, 0, '0', 'f', 'false', 'off']
```

**Comment by jemc** [comment]
My suggestion:

```ruby
def boolean(true_if: member([true, "true"]), false_if: member([false, "false"]))
  ->(input) do
    # returns true if the `true_if` coercion succeeds
    # otherwise, returns false if the `false_if` coercion succeeds
    # otherwise, an error
  end
end
```

This would allow @breckenedge (or any other user) to use their own custom set of collections, like this:

```ruby
boolean(
  true_if: member([true, 1, '1', 't', 'true', 'on']),
  false_if: member([false, 0, '0', 'f', 'false', 'off']),
)
```

This can nicely leverage the composability of these coercion functions, and give good flexibility for a user to be as loose or as strict as they want with their input.

One can also define their own convenience function that wraps this, so they don't need to redundantly define the set for every boolean in their application

```ruby
module MyCoerciveModule
  extend Coercive

  def booleanish
    boolean(
      true_if: member([true, 1, '1', 't', 'true', 'on']),
      false_if: member([false, 0, '0', 'f', 'false', 'off']),
    )
  end
end
```

**Comment by tonchis** [comment]
I think the collection you cast from should be configurable, because Coercive will never catch all the use cases. What about something like this?

```ruby
module Foo
  attribute : foo, boolean(map: {true => [1, "1", "t", "true"], false => [0, "0", "f", "false"]}), required
end

Foo.call("foo" => "t")
# => {"foo"=>true}
```

also, cc: @jemc, what do you think?

**Comment by jemc** [comment]
@tonchis - did you see my comment above suggesting a way to allow the collections to be configurable as nested coercion functions?

**Comment by tonchis** [comment]
@jemc ah! I didn't reload this page before posting my response. Yeah, that sounds good. I'll give it a shot

**Comment by tonchis** [comment]
Just published v1.6.0 with this feature! Closing here, thank you both!
